chainio/clients/txsender/privateKeyTxSender.go

```diff
@@ -0,0 +1,180 @@
+package txsender
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/Layr-Labs/eigensdk-go/chainio/clients/eth"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"github.com/Layr-Labs/eigensdk-go/signerv2"
+	sdktypes "github.com/Layr-Labs/eigensdk-go/types"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+var (
+	FallbackGasTipCap = big.NewInt(15_000_000_000)
+)
+
+var _ TxSender = (*privateKeyTxSender)(nil)
+
+type privateKeyTxSender struct {
+	ethClient eth.EthClient
+	address   common.Address
+	signerFn  signerv2.SignerFn
+	logger    logging.Logger
+
+	// cache
+	contracts map[common.Address]*bind.BoundContract
+}
+
+func NewPrivateKeyTxSender(ethRPCURL string, chainID *big.Int, ecdsaPrivateKey *ecdsa.PrivateKey, logger logging.Logger) (TxSender, error) {
+	ethClient, err := eth.NewClient(ethRPCURL)
+	if err != nil {
+		return nil, err
+	}
+	signer, addr, err := signerv2.SignerFromConfig(signerv2.Config{PrivateKey: ecdsaPrivateKey}, chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &privateKeyTxSender{
+		ethClient: ethClient,
+		address:   addr,
+		signerFn:  signer,
+		logger:    logger,
+	}, nil
+}
+
+func (t *privateKeyTxSender) SendTransaction(ctx context.Context, tx *types.Transaction) (TxID, error) {
+	// Estimate gas and nonce
+	// can't print tx hash in logs because the tx changes below when we complete and sign it
+	// so the txHash is meaningless at this point
+	t.logger.Debug("Estimating gas and nonce")
+	tx, err := t.estimateGasAndNonce(ctx, tx)
+	if err != nil {
+		return "", err
+	}
+
+	t.logger.Debug("Getting signer for tx")
+	signer, err := t.signerFn(ctx, t.address)
+	if err != nil {
+		return "", err
+	}
+
+	t.logger.Debug("Sending transaction")
+	opts := &bind.TransactOpts{
+		From:      t.address,
+		Nonce:     new(big.Int).SetUint64(tx.Nonce()),
+		Signer:    signer,
+		Value:     tx.Value(),
+		GasFeeCap: tx.GasFeeCap(),
+		GasTipCap: tx.GasTipCap(),
+		GasLimit:  tx.Gas(),
+		Context:   ctx,
+	}
+
+	contract := t.contracts[*tx.To()]
+	// if the contract has not been cached
+	if contract == nil {
+		// create a dummy bound contract tied to the `to` address of the transaction
+		contract = bind.NewBoundContract(*tx.To(), abi.ABI{}, t.ethClient, t.ethClient, t.ethClient)
+		// cache the contract for later use
+		t.contracts[*tx.To()] = contract
+	}
+
+	tx, err = contract.RawTransact(opts, tx.Data())
+	if err != nil {
+		return "", sdktypes.WrapError(fmt.Errorf("send: tx %v failed.", tx.Hash().String()), err)
+	}
+
+	return "", nil
```

Comment:
why are you returning "" here? Won't this not work with the waitForTransactionReceipt logic?

Reply:
guessing it should be `tx.Hash()` or something?

Reply:
ah yes! thanks for catching this